tool/src/org/antlr/v4/semantics/SymbolChecks.java

```diff
@@ -128,54 +131,102 @@ public void checkForTokenConflicts(List<GrammarAST> tokenIDRefs) {
-				// prior to the left recursion elimination step.
-				continue;
-			}
 
-			Map<String, LabelElementPair> labelNameSpace =
-					new HashMap<String, LabelElementPair>();
+			Map<String, LabelElementPair> labelNameSpace = new HashMap<>();
 			for (int i = 1; i <= r.numberOfAlts; i++) {
-				if (r.hasAltSpecificContexts()) {
-					labelNameSpace.clear();
-				}
-
 				Alternative a = r.alt[i];
 				for (List<LabelElementPair> pairs : a.labelDefs.values()) {
-					for (LabelElementPair p : pairs) {
-						checkForLabelConflict(r, p.label);
-						String name = p.label.getText();
-						LabelElementPair prev = labelNameSpace.get(name);
-						if (prev == null) labelNameSpace.put(name, p);
-						else checkForTypeMismatch(prev, p);
+					if (r.hasAltSpecificContexts()) {
+						// Collect labelName-labeledRules map for rule with alternative labels.
+						Map<String, List<LabelElementPair>> labelPairs = new HashMap<>();
+						for (LabelElementPair p : pairs) {
+							String labelName = findAltLabelName(p.label);
+							List<LabelElementPair> list;
+							if (labelPairs.containsKey(labelName)) {
+								list = labelPairs.get(labelName);
+							}
+							else {
+								list = new ArrayList<>();
+								labelPairs.put(labelName, list);
+							}
+							list.add(p);
+						}
+
+						for (List<LabelElementPair> internalPairs : labelPairs.values()) {
+							labelNameSpace.clear();
+							checkLabelPairs(r, labelNameSpace, internalPairs);
+						}
+					}
+					else {
+						checkLabelPairs(r, labelNameSpace, pairs);
 					}
 				}
 			}
 		}
 	}
 
-    void checkForTypeMismatch(LabelElementPair prevLabelPair, LabelElementPair labelPair) {
+	private void checkLabelPairs(Rule r, Map<String, LabelElementPair> labelNameSpace, List<LabelElementPair> pairs) {
+		for (LabelElementPair p : pairs) {
+			checkForLabelConflict(r, p.label);
+			String name = p.label.getText();
+			LabelElementPair prev = labelNameSpace.get(name);
+			if (prev == null) {
+				labelNameSpace.put(name, p);
+			}
+			else {
+				checkForTypeMismatch(r, prev, p);
+			}
+		}
+	}
+
+	private String findAltLabelName(CommonTree label) {
+		if (label == null) {
+			return null;
+		}
+		else if (label instanceof AltAST) {
+			AltAST altAST = (AltAST) label;
+			if (altAST.altLabel != null) {
+				return altAST.altLabel.toString();
+			}
+			else if (altAST.leftRecursiveAltInfo != null) {
+				return altAST.leftRecursiveAltInfo.altLabel.toString();
+			}
+			else {
+				return findAltLabelName(label.parent);
+			}
+		}
+		else {
+			return findAltLabelName(label.parent);
+		}
+	}
+
+	private void checkForTypeMismatch(Rule r, LabelElementPair prevLabelPair, LabelElementPair labelPair) {
 		// label already defined; if same type, no problem
 		if (prevLabelPair.type != labelPair.type) {
-			String typeMismatchExpr = labelPair.type + "!=" + prevLabelPair.type;
+			// TODO: replace rule token for left-recursive rule with correct token.
```

Comment:
I guess this is related to the left recursive rules transformation. Transformed rules have other token positions. Unfortunately I have no idea how to fix it easily.